rfc/0077-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,183 @@
+# Meta
+
+| Field          | Value                    |
+|----------------|--------------------------|
+| RFC Name       | Zone-Aware Replica Reads |
+| RFC ID         | 77                       |
+| Start Date     | 2024-04-25               |
+| Owner          | Sergey Avseyev           |
+| Current Status | DRAFT                    |
+| Revision       | #1                       |
+
+# Summary
+
+This document describes changes that SDKs must implement to optimize costs of
+network traffic when the server is being deployed across multiple availability
+zones.
+
+# Motivation
+
+Existing SDKs do not take into account node availability groups of the nodes in
+the cluster. In particular, when performing `getAllReplicas` operations, the SDK
+would send `GET (0x00)` request to active vBucket and `GET_REPLICA (0x83)` to
+each of the replicas. This approach will make reading from replicas expensive
+operations if nodes of the cluster deployed in different availability zones
+(AZ).
+
+# Use Cases
+
+## No Preference
+
+This is current behaviour, where the SDK does not take into account server
+groups.
+
+![figure 1: No Prefrence](figures/0077-case-1-no-preference.svg)
+
+## Local Only
+
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas.
+
+![figure 2: Local Only](figures/0077-case-2-local-only.svg)
+
+## Local First
+
+This case is a compromise between cost and availability, which trades speed. The
+idea is that the SDK will partition set of `[active, replica1, ...]` nodes into
+two groups: local and non-local. Then it will perform requests to local only,
+and only when neither of nodes will return data, send requests to non-local
```

Comment:
The trouble with LOCAL_FIRST is that if the non-local node doesn't return anything, it's usually going to be due to a timeout - network blip or undiscovered node failover.  So we only find out about it at the point the 2.5 KV user timeout has elapsed - and it's too late for us to try a fallback.

Instead of LOCAL_FIRST, we could support: The user can do an initial call with LOCAL_ONLY and a short timeout, and if that times out, try any fallback strategy they want.

Reply:
This strategy got simplified, the SDK does not do cascaded requests. It just filters the replica set according to the preference rules